pkg/resource/workspace/hooks.go

```diff
@@ -0,0 +1,220 @@
+)
+
+// workspaceCreating returns true if the supplied workspace is in the process
+// of being created
+func workspaceCreating(r *resource) bool {
+	if r.ko.Status.Status == nil {
+		return false
+	}
+	ws := *r.ko.Status.Status.StatusCode
+	return ws == string(svcapitypes.WorkspaceStatusCode_CREATING)
+}
+
+// workspaceCreating returns true if the supplied workspace is in an active state
+func workspaceActive(r *resource) bool {
+	if r.ko.Status.Status == nil {
+		return false
+	}
+	ws := *r.ko.Status.Status.StatusCode
+	return ws == string(svcapitypes.WorkspaceStatusCode_ACTIVE)
+}
+
+// customUpdateWorkspace patches each of the resource properties in the backend AWS
+// service API and returns a new resource with updated fields.
+func (rm *resourceManager) customUpdateWorkspace(
+	ctx context.Context,
+	desired *resource,
+	latest *resource,
+	delta *ackcompare.Delta,
+) (*resource, error) {
+
+	var err error
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.customUpdateWorkspace")
+	defer exit(err)
+
+	// Merge in the information we read from the API call above to the copy of
+	// the original Kubernetes object we passed to the function
+	ko := desired.ko.DeepCopy()
+
+	rm.setStatusDefaults(ko)
+
+	// Check if the state is active before updating
+	if !workspaceActive(latest) {
+		msg := "Cannot update workspace as it is not active, current status=" + string(*latest.ko.Status.Status.StatusCode)
+		ackcondition.SetSynced(desired, corev1.ConditionFalse, &msg, nil)
+		return desired, ackrequeue.NeededAfter(
+			errors.New(msg),
+			ackrequeue.DefaultRequeueAfterDuration,
+		)
+	}
+
+	if delta.DifferentAt("Spec.Tags") {
+		err = rm.updateWorkspaceTags(ctx, latest, desired)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if delta.DifferentAt("Spec.Alias") {
+		err = rm.updateWorkspaceAlias(ctx, desired)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &resource{ko}, nil
+}
+
+// updateWorkspaceTags uses TagResource and UntagResource to add, remove and update
+// workspace tags.
+func (rm *resourceManager) updateWorkspaceTags(
+	ctx context.Context,
+	latest *resource,
+	desired *resource,
+) error {
+	var err error
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.updateWorkspaceTags")
+	defer exit(err)
+
+	added, removed, updated := compareMaps(latest.ko.Spec.Tags, desired.ko.Spec.Tags)
+
+	if len(removed) > 0 {
+		removeTags := []*string{}
+		for i := range removed {
+			removeTags = append(removeTags, &removed[i])
+		}
+
+		input := &svcsdk.UntagResourceInput{
+			ResourceArn: (*string)(desired.ko.Status.ACKResourceMetadata.ARN),
+			TagKeys:     removeTags,
+		}
+
+		_, err = rm.sdkapi.UntagResourceWithContext(ctx, input)
+		rm.metrics.RecordAPICall("UPDATE", "UntagResource", err)
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(updated)+len(added) > 0 {
+		addedTags := map[string]*string{}
+		for k, v := range added {
+			addedTags[k] = v
+		}
+		for k, v := range updated {
+			addedTags[k] = v
+		}
```

Comment:
see comment below... you don't need to return added, removed, updated from `compareMaps`. Just return two parameters.